File: extensions/dirxmldev-visual/src/test/fishbone.test.ts
import * as assert from "node:assert/strict";
import * as fs from "node:fs";
import * as path from "node:path";
import { describe, it } from "node:test";
import { findTreeRoot } from "../ascode";
import { filesForNode, PUBLISHER_RIBS, RESOURCE_SETS, SPINE_SETS, SUBSCRIBER_RIBS } from "../fishbone";
import { fishboneGet, fishboneRefresh, fishboneReveal } from "../hooks";
import { findIdm } from "../idm";
import { fishboneHtml, fishboneLayout } from "../webview";

/**
 * Every model here comes from `bin/idm query … fishbone --json`, the one reader of
 * the manifests: these tests need the DirXMLDev checkout built (the launcher is
 * found upwards from sample-tree).
 */
const sample = path.resolve(__dirname, "..", "..", "sample-tree");

describe("sample as-code tree", () => {
  it("is a driverset.xml tree and walk-from-policy finds it", () => {
    assert.equal(findTreeRoot(sample), sample);
    const policy = path.join(sample, "drivers", "AD Driver", "subscriber", "sub-etp_Scoping.policy.xml");
    assert.equal(findTreeRoot(policy), sample);
  });

  it("finds bin/idm above the sample tree", () => {
    const idm = findIdm(sample);
    assert.ok(idm.command.endsWith(path.join("bin", "idm")), idm.command);
    assert.ok(fs.existsSync(idm.command));
  });

  it("loads the AD Driver with PolicySet keys from the Java enum", async () => {
    const model = await fishboneGet(sample, "AD Driver");
    assert.equal(model.driverSet.name, "driverset1");
    assert.equal(model.driver.name, "AD Driver");
    assert.equal(model.driver.dir, "drivers/AD Driver");
    const keys = new Set(
      [...model.publisher, ...model.subscriber, ...model.spine, ...model.resources]
        .filter((b) => b.policies.length > 0)
        .map((b) => b.key),
    );
    for (const k of [
      "subscriber-event",
      "subscriber-matching",
      "subscriber-create",
      "subscriber-placement",
      "subscriber-command",
      "publisher-event",
      "publisher-matching",
      "publisher-create",
      "publisher-placement",
      "publisher-command",
      "schema-mapping",
      "input",
      "output",
      "ecmascript",
      "gcv",
      "startup",
    ]) {
      assert.ok(keys.has(k), "missing set " + k);
    }
  });
});

describe("fishbone model", () => {
  it("puts Designer ribs on both channels and resolves files", async () => {
    const model = await fishboneGet(sample, "AD Driver");
    assert.equal(model.driver.name, "AD Driver");
    assert.equal(model.publisher.length, PUBLISHER_RIBS.length);
    assert.equal(model.subscriber.length, SUBSCRIBER_RIBS.length);
    assert.deepEqual(model.publisher.map((b) => b.key), PUBLISHER_RIBS.map((b) => b.key));
    assert.deepEqual(model.subscriber.map((b) => [b.key, b.label, b.setId]), SUBSCRIBER_RIBS.map((b) => [b.key, b.label, b.id]));
    assert.equal(model.spine.map((b) => b.key).join(","), SPINE_SETS.map((b) => b.key).join(","));

    const etp = model.subscriber.find((b) => b.key === "subscriber-event")!;
    assert.ok(etp.policies.length >= 2, "library + channel policy");
    const scoping = etp.policies.find((p) => p.name === "sub-etp_Scoping");
    assert.ok(scoping && !scoping.unresolved);
    assert.equal(scoping!.file, "drivers/AD Driver/subscriber/sub-etp_Scoping.policy.xml");
    assert.ok(fs.existsSync(path.join(sample, scoping!.file!)));

    const lib = etp.policies.find((p) => p.ref.startsWith("library/"));
    assert.ok(lib && lib.file === "library/lib-common-event.policy.xml");

    const schema = model.spine.find((b) => b.key === "schema-mapping")!;
    assert.equal(schema.policies[0]!.file, "drivers/AD Driver/sch_Map.policy.xml");

    const files = filesForNode(model, scoping!.id);
    assert.equal(files[0]!.file, scoping!.file);
    const boneFiles = filesForNode(model, "bone:schema-mapping");
    assert.equal(boneFiles[0]!.ref, "drivers/AD Driver/sch_Map");
    assert.ok(model.filter?.file?.endsWith("driver-filter.xml"));
  });

  it("renders the classic bone labels in the SVG", async () => {
    const model = await fishboneGet(sample, "AD Driver");
    const html = fishboneHtml(model, "n", "default-src 'none'", "fishbone.css");
    for (const label of [
      "Publisher channel",
      "Subscriber channel",
      "Identity Vault",
      "Application",
      "Event Transformation",
      "Matching",
      "Creation",
      "Placement",
      "Command Transformation",
      "Schema Mapping",
      "Input Transformation",
      "Output Transformation",
      "Filter",
      "ECMAScript",
      "GCVs",
      "Startup",
      "Shutdown",
      "sub-etp_Scoping",
      "sch_Map",
      "NOVLADENTEX-Startup-InitEntitlementConfigurationResource".slice(0, 20) + "…",
    ]) {
      assert.ok(html.includes(label), "missing " + label);
    }
  });

  it("maps AD-style Startup linkage to a driver-level bone with policies", async () => {
    const model = await fishboneGet(sample, "AD Driver");
    assert.deepEqual(
      model.resources.map((b) => b.key),
      RESOURCE_SETS.map((s) => s.key),
    );
    const startup = model.resources.find((b) => b.key === "startup")!;
    assert.equal(startup.setId, 15);
    assert.equal(startup.label, "Startup");
    assert.equal(startup.channel, "driver");
    assert.equal(startup.policies.length, 1);
    assert.equal(startup.policies[0]!.name, "NOVLADENTEX-Startup-InitEntitlementConfigurationResource");
    assert.equal(
      startup.policies[0]!.file,
      "drivers/AD Driver/NOVLADENTEX-Startup-InitEntitlementConfigurationResource.policy.xml",
    );
  });

  it("stacks GCV chips vertically instead of a comma-separated run-on", async () => {
    const model = await fishboneGet(sample, "AD Driver");
    const gcv = model.resources.find((b) => b.key === "gcv")!;
    assert.ok(gcv.policies.length >= 2, "sample links several GCV objects");
    const html = fishboneHtml(model, "n", "default-src 'none'", "fishbone.css");
    for (const p of gcv.policies) {
      assert.ok(html.includes(`data-node="${p.id}"`), "chip for " + p.name);
    }
  });

  it("reveal hook maps a ref to a real file", async () => {
    const file = await fishboneReveal(sample, "AD Driver", "drivers/AD Driver/subscriber/sub-etp_Scoping");
    assert.equal(file, path.join(sample, "drivers", "AD Driver", "subscriber", "sub-etp_Scoping.policy.xml"));
    assert.ok(fs.existsSync(file));
  });

  it("refresh hook writes the sentinel", () => {
    const sent = fishboneRefresh(sample);
    assert.ok(fs.existsSync(sent));
    fs.rmSync(path.dirname(sent), { recursive: true, force: true });
  });
});

describe("linkage.unknown startup (pre-PolicySet-15 trees)", () => {
  it("ingests linkage.unknown #15 as Startup when there is no named set", async () => {
    const dir = treeWithUnknownStartup(false);
    try {
      const model = await fishboneGet(dir, "AD Driver");
      const bone = model.resources.find((b) => b.key === "startup")!;
      assert.equal(bone.policies.length, 1, "must not invent a second startup link");
      assert.equal(bone.policies[0]!.name, "NOVLADENTEX-Startup-InitEntitlementConfigurationResource");
      assert.equal(bone.policies[0]!.unresolved, false);
      const html = fishboneHtml(model, "n", "default-src 'none'", "fishbone.css");
      assert.ok(html.includes("NOVLADENTEX-Startup-InitEntitlementConfigurationResource".slice(0, 20) + "…"));
    } finally {
      fs.rmSync(dir, { recursive: true, force: true });
    }
  });

  it("does not duplicate a named startup link when linkage.unknown also names it", async () => {
    const dir = treeWithUnknownStartup(true);
    try {
      const model = await fishboneGet(dir, "AD Driver");
      assert.equal(model.resources.find((b) => b.key === "startup")!.policies.length, 1);
    } finally {
      fs.rmSync(dir, { recursive: true, force: true });
    }
  });
});

describe("resource-row layout", () => {
  it("places Filter/Startup/Shutdown below a tall subscriber Creation/Placement stack", async () => {
    const model = await fishboneGet(sample, "AD Driver");
    const create = model.subscriber.find((b) => b.key === "subscriber-create")!;
    const place = model.subscriber.find((b) => b.key === "subscriber-placement")!;
    for (let i = 0; i < 8; i++) {
      create.policies.push({
        id: `policy:subscriber-create:tall:${i}`,
        ref: `drivers/AD Driver/tall-create-${i}`,
        name: `tall-create-${i}`,
        order: 100 + i,
        kind: "policy",
        unresolved: true,
      });
      place.policies.push({
        id: `policy:subscriber-placement:tall:${i}`,
        ref: `drivers/AD Driver/tall-place-${i}`,
        name: `tall-place-${i}`,
        order: 100 + i,
        kind: "policy",
        unresolved: true,
      });
    }
    const layout = fishboneLayout(model);
    const html = fishboneHtml(model, "n", "default-src 'none'", "fishbone.css");
    const startup = firstRect(html, "bone:startup");
    const shutdown = firstRect(html, "bone:shutdown");
    const filter = firstRect(html, "config:driver-filter");
    assert.equal(startup.y, layout.resourceY);
    assert.equal(shutdown.y, layout.resourceY);
    assert.equal(filter.y, layout.resourceY);
    assert.ok(layout.resourceY > 548, "tall subscriber must push the resource row below the old fixed y=548");

    for (const p of [...create.policies, ...place.policies]) {
      const chip = firstRect(html, p.id);
      assert.ok(
        chip.y + chip.h <= startup.y - 4,
        `${p.name} chip bottom ${chip.y + chip.h} overlaps Startup tile y=${startup.y}`,
      );
    }
    const viewBox = html.match(/viewBox="0 0 \d+ ([0-9.]+)"/);
    assert.ok(viewBox);
    assert.equal(Number(viewBox![1]), layout.height);
    assert.ok(layout.height > layout.resourceY + 48);
  });
});

const STARTUP_UNKNOWN =
  "cn=NOVLADENTEX-Startup-InitEntitlementConfigurationResource,cn=AD Driver,cn=driverset1,o=system#0#15";

/** Copy of sample-tree whose Startup exists only as linkage.unknown (or both); bin/idm is found upwards from the copy's real parent, so the copy is made inside the checkout. */
function treeWithUnknownStartup(keepNamedSet: boolean): string {
  const scratch = path.resolve(__dirname, "..", "..", ".test-scratch");
  fs.mkdirSync(scratch, { recursive: true });
  const dir = fs.mkdtempSync(path.join(scratch, "idm-unknown-startup-"));
  fs.cpSync(sample, dir, { recursive: true });
  const p = path.join(dir, "drivers", "AD Driver", "driver.xml");
  let xml = fs.readFileSync(p, "utf8");
  if (!keepNamedSet) {
    xml = xml.replace(/\s*<set key="startup">[\s\S]*?<\/set>/, "");
  }
  xml = xml.replace(
    /(<driver\b[^>]*>)/,
    `$1\n  <meta key="linkage.unknown.0">${STARTUP_UNKNOWN}</meta>`,
  );
  fs.writeFileSync(p, xml);
  return dir;
}

function firstRect(html: string, dataNode: string): { x: number; y: number; w: number; h: number } {
  const i = html.indexOf(`data-node="${dataNode}"`);
  assert.ok(i >= 0, "missing node " + dataNode);
  const slice = html.slice(i, i + 1200);
  const m = slice.match(/<rect\s+x="([^"]+)"\s+y="([^"]+)"\s+width="([^"]+)"\s+height="([^"]+)"/);
  assert.ok(m, "no rect for " + dataNode);
  return { x: Number(m![1]), y: Number(m![2]), w: Number(m![3]), h: Number(m![4]) };
}

File: src/test/java/com/pointblue/dirxml/dev/edit/FishboneTest.java
package com.pointblue.dirxml.dev.edit;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import com.pointblue.dirxml.dev.ascode.AsCodeReader;
import com.pointblue.dirxml.dev.ascode.AsCodeWriter;
import com.pointblue.dirxml.dev.json.Json;
import com.pointblue.dirxml.dev.model.Driver;
import com.pointblue.dirxml.dev.model.DriverSet;
import com.pointblue.dirxml.dev.model.PolicyLink;
import com.pointblue.dirxml.dev.model.PolicySet;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;
import java.util.Map;
import org.junit.Test;

/** {@code query fishbone --json}: the extension's payload, from the model, files named as the writer names them. */
public class FishboneTest {

    /** The extension's own smoke tree (AD Driver), kept in the repo. */
    private static final Path SAMPLE = Paths.get("extensions/dirxmldev-visual/sample-tree");

    @SuppressWarnings("unchecked")
    @Test
    public void sampleTreeFishboneHasEveryBoneAndRealFiles() throws Exception {
        DriverSet ds = AsCodeReader.read(SAMPLE);
        Driver ad = ds.driver("AD Driver");
        Map<String, Object> m = Fishbone.model(ds, SAMPLE, ad);
        assertEquals(SAMPLE.toAbsolutePath().normalize().toString(), m.get("treeRoot"));
        assertEquals("driverset1", ((Map<String, Object>) m.get("driverSet")).get("name"));
        assertEquals("drivers/AD Driver", ((Map<String, Object>) m.get("driver")).get("dir"));
        List<Object> pub = (List<Object>) m.get("publisher");
        List<Object> sub = (List<Object>) m.get("subscriber");
        assertEquals(5, pub.size());
        assertEquals(5, sub.size());
        assertEquals(3, ((List<Object>) m.get("spine")).size());
        assertEquals(4, ((List<Object>) m.get("resources")).size());
        assertEquals("Event Transformation", ((Map<String, Object>) sub.get(0)).get("label"));
        assertEquals("bone:subscriber-event", ((Map<String, Object>) sub.get(0)).get("id"));
        assertEquals(4, ((Map<String, Object>) sub.get(0)).get("setId"));
        int policies = 0;
        for (String section : List.of("publisher", "subscriber", "spine", "resources")) {
            for (Object bo : (List<Object>) m.get(section)) {
                for (Object po : (List<Object>) ((Map<String, Object>) bo).get("policies")) {
                    Map<String, Object> p = (Map<String, Object>) po;
                    policies++;
                    assertFalse(p.toString(), (Boolean) p.get("unresolved"));
                    assertNotNull(p.toString(), p.get("file"));
                    assertTrue(p.get("file").toString(), Files.isRegularFile(SAMPLE.resolve(p.get("file").toString())));
                    assertTrue(p.get("id").toString().startsWith("policy:"));
                }
            }
        }
        assertTrue("the sample links policies", policies > 5);
        Map<String, Object> filter = (Map<String, Object>) m.get("filter");
        assertEquals("config:driver-filter", filter.get("id"));
        assertTrue(Files.isRegularFile(SAMPLE.resolve(filter.get("file").toString())));
        // the Startup bone carries the AD entitlement startup policy
        Map<String, Object> startup = (Map<String, Object>) ((List<Object>) m.get("resources")).get(2);
        assertEquals("startup", startup.get("key"));
        assertFalse(((List<Object>) startup.get("policies")).isEmpty());
        // round trip through JSON keeps the shape
        Map<String, Object> back = Json.asMap(Json.parse(Json.pretty(m)));
        assertEquals(m.keySet(), back.keySet());
        assertTrue(Fishbone.text(m).contains("Schema Mapping (schema-mapping)"));
    }

    @SuppressWarnings("unchecked")
    @Test
    public void unresolvedLinkIsMarkedAndFilesFollowTheWriter() throws Exception {
        DriverSet ds = AsCodeReader.read(SAMPLE);
        Driver ad = ds.driver("AD Driver");
        ad.links.add(new PolicyLink(PolicySet.SUB_EVENT, "drivers/AD Driver/subscriber/Nope", 99));
        Map<String, Object> m = Fishbone.model(ds, SAMPLE, ad);
        List<Object> sub = (List<Object>) ((Map<String, Object>) ((List<Object>) m.get("subscriber")).get(0)).get("policies");
        Map<String, Object> last = (Map<String, Object>) sub.get(sub.size() - 1);
        assertEquals("Nope", last.get("name"));
        assertEquals("unresolved", last.get("kind"));
        assertEquals(Boolean.TRUE, last.get("unresolved"));
        assertFalse(last.containsKey("file"));
        Map<String, String> files = AsCodeWriter.files(ds);
        for (Map.Entry<String, String> e : files.entrySet()) {
            assertTrue(e.toString(), Files.isRegularFile(SAMPLE.resolve(e.getValue())));
        }
        Map<String, Object> drivers = Fishbone.drivers(ds, SAMPLE);
        assertEquals("AD Driver", ((Map<String, Object>) ((List<Object>) drivers.get("drivers")).get(0)).get("name"));
    }
}
